plugins/backstage-plugin-flux/src/hooks/useToggleSuspendResource.ts

```diff
@@ -0,0 +1,153 @@
+import { CustomResourceMatcher } from '@backstage/plugin-kubernetes-common';
+import { useAsyncFn } from 'react-use';
+import { gvkFromKind } from '../objects';
+import { Deployment, Source } from '../components/helpers';
+
+export const pathForResource = (
+  name: string,
+  namespace: string,
+  gvk: CustomResourceMatcher,
+): string => {
+  const basePath = [
+    '/apis',
+    gvk.group,
+    gvk.apiVersion,
+    'namespaces',
+    namespace,
+    gvk.plural,
+    name,
+  ].join('/');
+
+  return basePath;
+};
+
+export function toggleSuspendsuspendRequest(
+  name: string,
+  namespace: string,
+  clusterName: string,
+  gvk: CustomResourceMatcher,
+  suspend: boolean,
+) {
+  return {
+    clusterName,
+    path: pathForResource(name, namespace, gvk),
+    init: {
+      method: 'PATCH',
+      headers: {
+        'Content-Type': 'application/merge-patch+json',
+      },
+      body: JSON.stringify({
+        spec: {
+          suspend,
+        },
+      }),
+    },
+  };
+}
+
+export function getRequest(
+  name: string,
+  namespace: string,
+  clusterName: string,
+  gvk: CustomResourceMatcher,
+) {
+  return {
+    clusterName,
+    path: pathForResource(name, namespace, gvk),
+  };
+}
+
+export async function requestToggleSuspendResource(
+  kubernetesApi: KubernetesApi,
+  name: string,
+  namespace: string,
+  clusterName: string,
+  gvk: CustomResourceMatcher,
+  suspend: boolean,
+) {
+  const res = await kubernetesApi.proxy(
+    toggleSuspendsuspendRequest(name, namespace, clusterName, gvk, suspend),
+  );
+  const key = suspend ? 'Suspend' : 'Resume';
+  if (!res.ok) {
+    throw new Error(
+      `Failed to ${key} resource: ${res.status} ${res.statusText}`,
+    );
+  }
+}
+
+export async function toggleSuspendResource(
+  resource: Source | Deployment,
+  kubernetesApi: KubernetesApi,
+  alertApi: AlertApi,
+  suspend: boolean,
+) {
+  const key = suspend ? 'Suspend' : 'Resume';
+  try {
+    const gvk = gvkFromKind(resource.type);
+    if (!gvk) {
+      throw new Error(`Unknown resource type: ${resource.type}`);
+    }
+
+    if ('sourceRef' in resource && resource.sourceRef) {
+      const sourceGVK = gvkFromKind(resource.sourceRef.kind);
+      if (!sourceGVK) {
+        throw new Error(
+          `Unknown resource source type: ${resource.sourceRef.kind}`,
+        );
+      }
+      // sync the source
+      await requestToggleSuspendResource(
+        kubernetesApi,
+        resource.sourceRef.name!,
+        resource.sourceRef.namespace || resource.namespace,
+        resource.clusterName,
+        sourceGVK,
+        suspend,
+      );
+    }
+
+    // sync the helm release
+    await requestToggleSuspendResource(
+      kubernetesApi,
+      resource.name,
+      resource.namespace,
+      resource.clusterName,
+      gvk,
+      suspend,
+    );
+
+    alertApi.post({
+      message: `${key} request successful`,
+      severity: 'success',
+      display: 'transient',
+    });
+  } catch (e: any) {
```

Comment:
Can we add the Error type here?

Reply:
typescript whine about it. It must be any or unknown.
https://stackoverflow.com/questions/69021040/why-catch-clause-variable-type-annotation-must-be-any